lib/cli/src/Cardano/CLI.hs

```diff
@@ -367,114 +377,139 @@ runCli = join . customExecParser preferences
 
 cmdKey :: Mod CommandFields (IO ())
 cmdKey = command "key" $ info (helper <*> cmds) $ mempty
-    <> progDesc "Derive keys from mnemonics."
+    <> progDesc "Derive and manipulate keys."
+    <> headerDoc (Just $ string $ mconcat
+        [ "Keys can be passed as arguments or read as standard input. Both "
+        , "bech32- and hexadecimal encodings are supported."
+        , "\n\n"
+        , "For instance:\n"
+        , "$ cardano-wallet-byron key root --wallet-style icarus --encoding bech32 -- express theme celery coral permit ... \\\n"
+        , "    | cardano-wallet-byron key public\n"
+        , "xpub1k365denpkmqhj9zj6qpax..."
+        ])
   where
     cmds = subparser $ mempty
         <> cmdKeyRoot
         <> cmdKeyChild
         <> cmdKeyPublic
         <> cmdKeyInspect
 
--- | Record with mnemonic and key derivation funcionality — /without/ any type
+-- | Record encapsulating @mnemonicToRootKey@  — /without/ any type
 -- parameters related to scheme.
 --
 -- This means that we can have a value for byron, a value for icarus, both with
 -- the same type @CliKeyScheme@.
 --
 -- @CliKeyScheme@ is on the other hand parameterized over @key@ and @m@.
--- @hoistKeyScheme@ is provided for mapping over @m@. @mapKey@ is provided for
--- mapping over @key@.
---
--- This way we can test mapping @XPrv@s to @Text@ as pure code in
--- @Either String@, rather than @IO@.
 data CliKeyScheme key m = CliKeyScheme
     { allowedWordLengths :: [Int]
     , mnemonicToRootKey :: [Text] -> m key
-    , deriveChildKey :: key -> DerivationIndex -> m key
-    , toPublic :: key -> m key
-    , inspect :: key -> m Text
     }
 
 data XPrvOrXPub = AXPrv XPrv | AXPub XPub
 
--- | Change the underlying monad of a @CliKeyScheme@.
-hoistKeyScheme
-    :: (forall a. m1 a -> m2 a)
-    -> CliKeyScheme key m1
-    -> CliKeyScheme key m2
-hoistKeyScheme eta s = CliKeyScheme
-    { allowedWordLengths = allowedWordLengths s
-    , mnemonicToRootKey = eta . mnemonicToRootKey s
-    , deriveChildKey = \k i -> eta $ deriveChildKey s k i
-    , toPublic = eta . toPublic s
-    , inspect = eta . inspect s
-    }
+eitherToIO :: Either String a -> IO a
+eitherToIO (Right a) = return a
+eitherToIO (Left e) = hPutStrLn stderr e >> exitFailure
 
--- | Pair of functions representing the bidirectional transformation between
--- a @XPrvOrXPub@ and its hex-encoded form.
-keyHexCodec :: (XPrvOrXPub -> Either String Text, Text -> Either String XPrvOrXPub)
-keyHexCodec = (encode, decode)
+data KeyEncoding = Hex | Bech32
+    deriving (Show, Eq, Enum, Bounded)
+
+detectEncoding :: Text -> Either String KeyEncoding
+detectEncoding x
+    | "xprv1" `T.isPrefixOf` x = pure Bech32
+    | "xpub1" `T.isPrefixOf` x = pure Bech32
```

Comment:
Detecting bech32 encoding by detecting the prefix is sort of wrong. A value could be bech32 encoded but simply has a wrong prefix. Yet, a value can't be both bech32 and hex encoded, so we may as well simply try to decode them, and then, check for the prefix in case it's bech32.

Reply:
The problem is that I want `valid_xprv1 <> "ö"` to fail with a `bech32` specific error message, but `0000` to fail with a hex-specific one.

Not sure how to achieve that otherwise.

I experimented some with:

```haskell
    let r1 = (,Hex) <$> decodeHex txt
    let r2 = (,Bech32) <$> decodeBech32 txt
    case (r1, r2) of
        (Right x, _) -> return x
        (_, Right x) -> return x
        (Left (Recognized e), Left (UnrecognizedEncoding _))
            -> Left e
        (Left (UnrecognizedEncoding _), Left (Recognized e))
            -> Left e
        (Left (UnrecognizedEncoding _), Left (UnrecognizedEncoding _))
            -> Left fullKeyEncodingDescription
        (Left (Recognized _), Left (Recognized _))
            -> Left "decodeAnyKey: internal error. The key is recognized as \
                    \bech32- and hex-encoded at the same time, but fails with \
                    \both."
```
but it didn't seem to work out fully.

Reply:
I added golden tests for failures in 1d0cc5effcf02321e0ef0f3d3069d54328d5c726